### PR TITLE
Fix up HibernateObserver to allow cascading persistence after bug report

### PR DESCRIPTION
**Pull request by rick-r3**
Fixes https://github.com/corda/corda/issues/514

Switch from `StatelessSession` to normal `Session` since stateless version will not cascade persistence, and adds unit test to ensure persistence cascades.

Have run the Explorer demo in simulation mode (which repeatedly persists cash) connected to YourKit and it shows no memory leaks involving Hibernate so I think I have successfully disabled any caching and any other state retention (which is the original motivation for using `StatelessSession` since we are write-only).

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/services/schema/HibernateObserver.kt [hunk 1]
```diff
@@ -87,14 +86,38 @@ class HibernateObserver(vaultService: VaultService, val schemaService: SchemaSer
...
-        val session = sessionFactory.openStatelessSession(TransactionManager.current().connection)
+        val session = sessionFactory.withOptions().
+                connection(TransactionManager.current().connection).
+                flushMode(FlushMode.MANUAL).
+                openSession()
         session.use {
             val mappedObject = schemaService.generateMappedObject(state, schema)
             mappedObject.stateRef = PersistentStateRef(stateRef)
-            session.insert(mappedObject)
+            session.persist(mappedObject)
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/schema/HibernateObserver.kt]: Use `it`

File: node/src/main/kotlin/net/corda/node/services/schema/HibernateObserver.kt [hunk 2]
```diff
@@ -87,14 +86,38 @@ class HibernateObserver(vaultService: VaultService, val schemaService: SchemaSer
...
+        metadataBuilder.applyPhysicalNamingStrategy(object : PhysicalNamingStrategyStandardImpl() {
+            override fun toPhysicalTableName(name: Identifier?, context: JdbcEnvironment?): Identifier {
+                val default = super.toPhysicalTableName(name, context)
+                return Identifier.toIdentifier(tablePrefix + default.text, default.isQuoted)
+            }
+        })
+
+        val metadata = metadataBuilder.build()
+
+        val sessionFactoryBuilder = metadata.getSessionFactoryBuilder()
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/schema/HibernateObserver.kt]: Property access

File: node/src/main/kotlin/net/corda/node/services/schema/HibernateObserver.kt [hunk 3]
```diff
@@ -87,14 +86,38 @@ class HibernateObserver(vaultService: VaultService, val schemaService: SchemaSer
...
+        sessionFactoryBuilder.allowOutOfTransactionUpdateOperations(true)
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/schema/HibernateObserver.kt]: You could put all this in a `run` block and avoid the need to assign `sessionFactoryBuilder`

File: node/src/test/kotlin/net/corda/node/services/HibernateObserverTests.kt [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        @ManyToOne(fetch = FetchType.LAZY)
+        @JoinColumns(JoinColumn(name = "transaction_id"), JoinColumn(name = "output_index"))
+        var parent: Parent? = null
+    }
+
+    @Test
+    fun `test child objects are persisted`() {
+        val testSchema = object : MappedSchema(SchemaFamily::class.java, 1, setOf(Parent::class.java, Child::class.java)) {}
+        val rawUpdatesPublisher = PublishSubject.create<Vault.Update>()
+        val vaultService = object : VaultService {
```
Comment (shamsasari) [inline node/src/test/kotlin/net/corda/node/services/HibernateObserverTests.kt]: I'd rather you passed in `VaultService.rawUpdates` to the HibernateObserver and avoid this ugliness.

File: node/src/test/kotlin/net/corda/node/services/HibernateObserverTests.kt [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+                parent.children.add(Child())
+                parent.children.add(Child())
+                return parent
+            }
+
+        }
+        @Suppress("UNUSED_VARIABLE")
+        val observer = HibernateObserver(vaultService, schemaService)
+        databaseTransaction(database) {
+            rawUpdatesPublisher.onNext(Vault.Update(emptySet(), setOf(StateAndRef(TransactionState(object : QueryableState {
```
Comment (shamsasari) [inline node/src/test/kotlin/net/corda/node/services/HibernateObserverTests.kt]: This QueryableState shouldn't be an anonymous class

**Comment by rick-r3** [comment]
@shamsasari Thanks for your comments.  Think I addressed them all.  Ready for review again.

File: node/src/test/kotlin/net/corda/node/services/HibernateObserverTests.kt [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        override val participants: List<CompositeKey>
+            get() = throw UnsupportedOperationException()
+    }
+
+    // This method does not use back quotes for a nice name since it seems to kill the kotlin compiler.
+    @Test
+    fun testChildObjectsArePersisted() {
+        val testSchema = object : MappedSchema(SchemaFamily::class.java, 1, setOf(Parent::class.java, Child::class.java)) {}
+        val rawUpdatesPublisher = PublishSubject.create<Vault.Update>()
+        val vaultService = object : VaultService {
```
Comment (shamsasari) [inline node/src/test/kotlin/net/corda/node/services/HibernateObserverTests.kt]: This still needs addressing

**Comment by rick-r3** [comment]
Addressed the comment I missed

**Pull request merged**